File: Data Structures & Algorithms/house-robber/submission-4.rs

impl Solution {
    pub fn rob(mut nums: Vec<i32>) -> i32
    {
        use std::cmp::max;

        if nums.len() == 0 { return 0; }
        if nums.len() == 1 { return nums[0]; }
        if nums.len() == 2 { return max(nums[0],nums[1]); }

        for i in 0..nums.len()
        {
            nums[i] = std::cmp::max(
                if i < 1 {0} else { nums[i-1] },
                if i < 2 {0} else { nums[i-2] } + nums[i],
            );
        }
        return nums.pop().unwrap_or(0);
    }
}

File: Data Structures & Algorithms/house-robber-ii/submission-4.cpp

class Solution {
public:
    int rob(vector<int>& nums)
    {
        auto constexpr rob_linear = [](auto iter, auto const& end)
        {
            auto dp = array<int,3>{};

            for (; iter != end; ++iter) {
                dp = { dp[1], dp[2], *iter + max(dp[0], dp[1]) };
            }
            return max(dp[1], dp[2]);
        };

        switch (nums.size()) {
            break; case 0: return 0;
            break; case 1: return nums[0];
            break; default: return max(
                rob_linear(nums.cbegin(), std::prev(nums.cend())),
                rob_linear(std::next(nums.cbegin()), nums.cend())
            );
        }
    }
};
